### PR TITLE
Use `as` instead of `url` to determine URL newness

### LINKED ISSUE #3065: getInitialProps not working with history (back button)?
I have an app with a single page `index.js`, and I parse the query parameters in `getInitialProps`:

	static async getInitialProps ({query}) {
		console.log('PAGE', query.screenId);
		try {
			const projectData = await ProjectLoader.loadProject(query);
			return projectData;
		}
		catch (err) {
			console.error(err);
		}
	};

I use `<Link href={pathname, query} prefetch>` tags to build my links.

The `screenId` property is key to finding the right page to render.

However, when I use the back button in the browser to a page without `screenId` set, `getInitialProps` doesn’t return `screenId:undefined` as expected, but another value (as you can see in the console log):

[![getInitialProps not working with history][1]][1]

When I refresh the page, I get the right properties and see the correct results.

The “imaginary” `screenId` is actually a valid value, but I have no idea where this value comes from.

Why is this happening?

  [1]: https://i.stack.imgur.com/pWUXE.gif

### PR DESCRIPTION
**Pull request by gcpantazis**
Currently, using `as` will cause the router to think the URL is not changing in the case where you're re-rendering the same page with a different route. This would most likely be an issue for custom servers
which are using shallow routing.

This should be an invisible change for non-custom-server users, since `as` is defaulted to `url` if not set.

This should resolve #3065.

### CODE REVIEW

File: lib/router/router.js [hunk 1]
```diff
@@ -153,13 +153,14 @@ export default class Router {
       return true
     }
 
+    const { asPathname, asQuery } = parse(as, true)
```
Comment (bathos) [inline lib/router/router.js]: doesn’t this need to be `const { pathname: asPathname, query: asQuery } = parse(as, true)`?

Comment (gcpantazis) [reply lib/router/router.js]: Woo, yes. 🤦‍♂️  Let me double check that this is working correctly.

Comment (gcpantazis) [reply lib/router/router.js]: Yeah looks good. Patched.

**Comment by bathos** [comment]
I’m very excited to see this getting fixed. We currently have to patch Router.prototype in our app to fix it and ... well, it was a precision surgery, but I’m pretty sure I could be sent to prison in like 37 countries for how I did it. Once this lands I’ll be able to sleep at nite again. Thanks @gcpantazis!

**Comment by SBoudrias** [comment]
Looks like this branch is out of sync with master now. Is there still plans to get this bug fixed?

**Review by timneutkens — APPROVED** [review]

**Pull request merged**